### DISCUSSION TITLE
Use custom fields as valueSource

### DISCUSSION THREAD

**Discussion by memoricab** [react-querybuilder/react-querybuilder#501] [Q&A] [answered]
Hello,

I enable field as valueSources for comparison. But fields are same with the left side of the equation, I would like to use custom set of fields.

Note, I still want to provide bot valueSources field or value.

Is there a way to achieve this?

Thanks

**Comment by jakeboone02** [comment, marked as answer]
I would suggest intercepting/overriding the `values` prop on the value editor component, but only when `valueSource` is `"field"`.

Working example here:  https://codesandbox.io/p/sandbox/vibrant-fire-89tplm

Relevant code:

```tsx
import { Field, ValueEditor, ValueEditorProps } from 'react-querybuilder';

const otherFields: Field[] = [
  { name: 'some', label: 'Some' },
  { name: 'other', label: 'Other' },
  { name: 'fields', label: 'Fields' },
];

export const CustomValueEditor = (props: ValueEditorProps) => {
  if (props.valueSource === 'field') {
    return <ValueEditor {...props} values={otherFields} />;
  }

  // Fall back to default value editor when valueSource is "value"
  return <ValueEditor {...props} />;
};
```

Implementation:

```tsx
<QueryBuilder
  fields={fields}
  query={query}
  onQueryChange={q => setQuery(q)}
  getValueSources={() => ['value', 'field']}
  controlElements={{
    valueEditor: CustomValueEditor,
  }}
/>
```

**Reply by memoricab** [reply]
this is a MACHINE-like answer, thanks @jakeboone02

**Reply by jakeboone02** [reply]
Glad to help!
